pydantic_ai_slim/pydantic_ai/prompt_config.py

```diff
@@ -0,0 +1,414 @@
+    """
+
+    validation_errors_retry_prompt: str | Callable[[RetryPromptPart, RunContext[Any]], str] | None = None
+    """Message appended to validation errors when asking the model to retry.
+
+    If `None`, uses the default: 'Fix the errors and try again.'
+    """
+
+    tool_retry_prompt: str | Callable[[RetryPromptPart, RunContext[Any]], str] | None = None
+    """Message appended when a `ModelRetry` exception is raised from a tool.
+
+    If `None`, uses the default: 'Fix the errors and try again.'
+    """
+
+    no_tool_retry_prompt: str | Callable[[RetryPromptPart, RunContext[Any]], str] | None = None
+    """Message appended when a `ModelRetry` exception is raised outside of a tool context.
+
+    If `None`, uses the default: 'Fix the errors and try again.'
+    """
+
+    prompted_output_template: str | None = None
+    """Template for prompted output schema instructions.
+
+    If set, this overrides the template used for [`PromptedOutput`][pydantic_ai.output.PromptedOutput].
+
+    If `None`, the template from `PromptedOutput` is used if set, otherwise the model profile default
+    ([`ModelProfile.prompted_output_template`][pydantic_ai.profiles.ModelProfile.prompted_output_template])
+    is used.
+    """
+
+    description_template: Callable[[str | list[pydantic_core.ErrorDetails]], str] | None = None
+    """Format a description message while asking the model to retry."""
+
+    def apply_template(self, message_part: ModelRequestPart, ctx: RunContext[Any]) -> ModelRequestPart:
+        if isinstance(message_part, ToolReturnPart):
+            if message_part.return_kind in (None, 'tool-executed'):
+                return message_part
+
+            field_name = message_part.return_kind.replace('-', '_')
+            template = getattr(self, field_name, None)
+            # Map return_kind directly to template attribute name (e.g. 'final-result-processed' -> 'final_result_processed')
+
+            # Special case for tool-denied: only apply if template is explicitly set
+            if message_part.return_kind == 'tool-denied':
+                return self._apply_tool_template(message_part, ctx, template) if template else message_part
+
+            if template := template or return_kind_to_default_prompt_template.get(message_part.return_kind):
+                return self._apply_tool_template(message_part, ctx, template)
+
+        elif isinstance(message_part, RetryPromptPart):
+            return self._apply_retry_template(message_part, ctx)
+
+        return message_part
+
+    def apply_template_message_history(self, _messages: list[ModelMessage], ctx: RunContext[Any]) -> list[ModelMessage]:
+        return [
+            replace(
+                message,
+                parts=[self.apply_template(part, ctx) for part in message.parts],
+            )
+            if isinstance(message, ModelRequest)
+            else message
+            for message in _messages
+        ]
+
+    def _apply_retry_template(
```

Comment:
This method is now mirroring `RetryPromptPart.model_response()` in terms of the logic; can we unify them somehow and only have the logic in one place?

Maybe `model_response()` could take `prompt_config: PromptConfig = DEFAULT_PROMPT_CONFIG` as an arg and keep the logic there, or we could have the logic as a method on `PromptConfig` and have `model_response` call `return self.retry_message or DEFAULT_PROMPT_CONFIG.format_retry_prompt(RetryPromptPart)`

Reply:
Yeah I don't like this either, I'll change to keep it in one place. We wouldn't want to maintain same code in two files.